src/referenced-fields.ts

```diff
@@ -294,3 +313,271 @@ export const getModifiedFields: EngineInstance['getModifiedFields'] = (
 		return checkQuery(abstractSqlQuery);
 	}
 };
+
+type AliasedTable = {
+	tableName: string;
+	alias: string;
+};
+
+type BindCandidate = AliasedTable & {
+	scope: AbstractSqlQuery;
+};
+
+const findTablesInConstraint = (
+	constraint: BooleanTypeNodes,
+	candidates: BindCandidate[],
+) => {
+	// Recurse until we're sure that there are no more subqueries
+	switch (constraint[0]) {
+		case 'Boolean':
+			break;
+		case 'Equals':
+		case 'NotEquals':
+		case 'GreaterThan':
+		case 'GreaterThanOrEqual':
+		case 'LessThan':
+		case 'LessThanOrEqual':
+		case 'IsDistinctFrom':
+		case 'IsNotDistinctFrom':
+			const equalsNode = constraint as EqualsNode;
+			const left = equalsNode[1];
+			const right = equalsNode[2];
+			if (!isAbstractSqlQuery(left) || !isAbstractSqlQuery(right)) {
+				throw new Error(
+					'cannot introspect into the string form of AbstractSqlQuery',
+				);
+			}
+			if (SELECT_QUERY_TYPES.includes(left[0])) {
+				findBindCandidates(left, candidates);
+			}
+			if (SELECT_QUERY_TYPES.includes(right[0])) {
+				findBindCandidates(right, candidates);
+			}
+			break;
+		case 'In':
+		case 'NotIn':
+			const inNode = constraint as InNode;
+			for (const arg of inNode.slice(2)) {
+				if (!isAbstractSqlQuery(arg)) {
+					throw new Error(
+						'cannot introspect into the string form of AbstractSqlQuery',
+					);
+				}
+				if (SELECT_QUERY_TYPES.includes(arg[0])) {
+					findBindCandidates(arg, candidates);
+				}
+			}
+			break;
+		case 'Exists':
+		case 'NotExists':
+			const existsNode = constraint as ExistsNode;
+			const inner = existsNode[1];
+			if (!isAbstractSqlQuery(inner)) {
+				throw new Error(
+					'cannot introspect into the string form of AbstractSqlQuery',
+				);
+			}
+			if (SELECT_QUERY_TYPES.includes(inner[0])) {
+				findBindCandidates(inner, candidates);
+			}
+			break;
+		case 'Not':
+			const notNode = constraint as NotNode;
+			findTablesInConstraint(notNode[1], candidates);
+			break;
+		case 'And':
+		case 'Or':
+			const andNode = constraint as AndNode;
+			for (const arg of andNode.slice(1)) {
+				// TODO: type checking is going awry here for some reason
+				findTablesInConstraint(arg as any, candidates);
+			}
```

Comment:
It might work if you did something like
```suggestion
			const [ , ...args] = andNode;
			for (const arg of args) {
				findTablesInConstraint(arg, candidates);
			}
```
as I expect the issue is that `.slice` returns `(typeof andNode)[number]` and doesn't play nice with tuples

Reply:
I just fixed this. It was missing an `!isAbstractSqlQuery` check which does the coercion.